sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CertificateValidationCheck.cs

```diff
@@ -0,0 +1,299 @@
+        }
+
+        private static ImmutableArray<Location> InvocationLocations(InspectionContext c, MethodDeclarationSyntax method)
+        {
+            var returnExpressionSublocationsList = method.Body.DescendantNodes()
+                .OfType<ReturnStatementSyntax>()
+                .Select(x => x.Expression)
+                .Where(x => !IsVisited(c, x));      //Ignore all return statements with recursive call. Result depends on returns that could return compliant validator.
+            return MultiExpressionSublocations(c, returnExpressionSublocationsList);
+        }
+
+        private static bool IsVisited(InspectionContext c, ExpressionSyntax expression)
+        {
+            if (expression is InvocationExpressionSyntax invocation)
+            {
+                var symbol = c.Context.SemanticModel.GetSymbolInfo(invocation).Symbol;
+                return symbol != null && !symbol.DeclaringSyntaxReferences.IsEmpty
+                    && symbol.DeclaringSyntaxReferences.Select(x => x.GetSyntax() as MethodDeclarationSyntax).Any(x => x != null && c.VisitedMethods.Contains(x));
+            }
+            return false;
+        }
+
+        private static ImmutableArray<Location> MultiExpressionSublocations(InspectionContext c, IEnumerable<ExpressionSyntax> expressions)
+        {
+            var exprSublocationsList = expressions.Distinct(new Helpers.CSharp.CSharpSyntaxNodeEqualityComparer<ExpressionSyntax>())
+                .Select(x => CallStackSublocations(c, x))
+                .ToArray();
+            if (exprSublocationsList.Any(x => x.IsEmpty))   //If there's at leat one concurrent expression, that returns compliant delegate, than there's some logic and this scope is compliant
+            {
+                return ImmutableArray<Location>.Empty;
+            }
+            return exprSublocationsList.SelectMany(x => x).ToImmutableArray();      //Else every return statement is noncompliant
+        }
+
+        private static ImmutableArray<Location> CallStackSublocations(InspectionContext c, ExpressionSyntax expression)
+        {
+            var lst = ArgumentLocations(c, expression);
+            if (!lst.IsEmpty)        //There's noncompliant issue in this chain
+            {
+                var Loc = expression.GetLocation();
+                if (!lst.Any(x => x.SourceSpan.IntersectsWith(Loc.SourceSpan)))
+                {   //Add 2nd, 3rd, 4th etc //Secondary marker. If it is not marked already from direct Delegate name or direct Lambda occurance
+                    return lst.Concat(new[] { Loc }).ToImmutableArray();
+                }
+            }
+            return lst;
+        }
+
+        private static ImmutableArray<Location> BlockLocations(BlockSyntax block)
+        {
+            var ret = ImmutableArray.CreateBuilder<Location>();
+            if (block != null)
+            {
+                //ToDo: VB.NET vs. return by assign to function name
```

Comment:
we need to get alligned between `TODO` and `ToDo` - normally in the codebase it's `TODO` :)

Reply:
no need to change now, just to keep in mind for the coding style discussion